src/Traits/FiltersQueries.php

```diff
@@ -32,28 +37,38 @@ protected function applyFilters(Builder $query, array $args): Builder
      * Apply filters recursively.
      *
      * @param \Illuminate\Database\Eloquent\Builder $query
-     * @param array $args
+     * @param Arguments $args
      * @param mixed $type
      * @return \Illuminate\Database\Eloquent\Builder
      */
-    protected function applyFiltersRecursively(Builder $query, array $args, $type = null): Builder
+    protected function applyFiltersRecursively(Builder $query, Arguments $args = null, $type = null): Builder
     {
         foreach ($args as $key => $value) {
             if ($key === 'AND' || $key === 'OR') {
                 $query->where(function ($query) use ($value, $key) {
                     foreach ($value as $set) {
-                        if (! empty($set)) {
-                            $this->applyFiltersRecursively($query, $set ?? [], $key);
+                        if ( ! empty($set)) {
+                            $this->applyFiltersRecursively($query, $set, $key);
                         }
                     }
                 });
             } else {
                 $schema = $this->registry->resolveSchemaForModel(get_class($query->getModel()));
 
                 if ($schema->getRelationFields()->has($key)) {
-                    $this->applyRelationFilter($query, $key, $value, $type);
+                    // TODO: Extract this.
+                    $relation = $schema->getRelationFields()->first(function (Field $field, string $fieldKey) use ($key) {
+                        return $key === $fieldKey;
+                    })->getAccessor();
+
+                    $this->applyRelationFilter($query, $relation, $value, $type);
                 } else {
-                    $this->filter($query, $key, $value, $type);
+                    // TODO: Extract this.
+                    $column = $schema->getFields()->first(function (Field $field, string $fieldKey) use ($key) {
+                        return $key === $fieldKey;
+                    })->getAccessor();
```

Comment:
This fails for dynamic keys (e.g. `name_starts_with`) so we need to find a way to extract the key from that.